### PR TITLE
feat(cli): Support multiple configuration files.

### PR DESCRIPTION
**Pull request by Jeffail**
This currently works by parsing each discrete file individually, and appending them to a root config. This has the advantage of being able to report the specific file that fails to parse or introduces a duplicate. However, it also means we need to manually determine how parsed config structs are to be merged.

Global fields may be specified multiple times, but if there is a collision then it is considered an error. This makes it possible to run configs together that also work individually.

### CODE REVIEW

**Comment by binarylogic** [comment]
Nice!

> Global fields may be specified multiple times, but if there is a collision then it is considered an error. This makes it possible to run configs together that also work individually.

A couple of thoughts:

1. We could load all files in order (maybe alphabetically)
2. We could make no guarantee on the order. Whatever happens, happens.

I lean towards 2, and we can encourage users to only define options once. If they don't we make no guarantee on the order or priority.

**Comment by jszwedko** [comment]
> I lean towards 2, and we can encourage users to only define options once. If they don't we make no guarantee on the order or priority.

Personally, I'd prefer an explicit error about key conflict so that I can fix the configuration rather than silently depending on an undefined ordering as the failure cases are likely to cause anguish:

* I accidentally define a key twice and are implicitly depending on the ordering of however Vector chooses to load them and this ordering changes between versions
* I go to add a key and wonder why it isn't being applied (because it is defined somewhere else)

**Comment by binarylogic** [comment]
@jszwedko I want to solve the world climate crisis, beat Trump, and own a mac pro, but I can't have everything. 😄 

Joking aside, these are good points. We should have an explicit ordering. @Jeffail how difficult would it be to warn on conflicts?

**Comment by Jeffail** [comment]
Right now we get:

- No configs set a global: we use default
- One config sets a global: we use that value
- Multiple configs set a global to the same value: use that value
- Multiple configs set a global to different values: we throw an error

Which I think is the best we can do. It allows you to have configs `foo.toml` and `bar.toml`, run them independently with a global field set (consistently), and then run them together without aggro. If you really need your individual configs to run with different values for things like `data_dir` then you can have separate configs `globs_1.toml`, `globs_2.toml` and run `foo.toml`, `bar.toml`, or both with whichever globals you want at the time:

```sh
vector -c ./globs_1.toml -c ./foo.toml
vector -c ./globs_2.toml -c ./bar.toml
vector -c ./globs_1.toml -c ./foo.toml
```

File: src/main.rs [hunk 1]
```diff
@@ -220,23 +217,35 @@ fn main() {
...
-        file
+    let mut config_paths = Vec::new();
+    for config_pattern in if !opts.config_paths.is_empty() {
+        opts.config_paths
     } else {
-        std::process::exit(exitcode::CONFIG);
-    };
+        DEFAULT_CONFIG_PATHS.to_vec()
+    } {
+        let matches: Vec<PathBuf> = glob(config_pattern.to_str().expect("no ability to glob"))
```
Comment (LucioFranco) [inline src/main.rs]: These expect messages seem like they would be better to be error! logs? I would imagine we only want to panic on truly fatal errors.

File: src/main.rs [hunk 2]
```diff
@@ -415,6 +418,43 @@ fn handle_config_errors(config: Result<Config, Vec<String>>) -> Option<Config> {
...
+            return;
+        };
+
+        trace!(
+            message = "Parsing config.",
+            path = ?p
+        );
+
+        match Config::load(file) {
+            Ok(next_config) => match config.append(next_config) {
```
Comment (LucioFranco) [inline src/main.rs]: minor nit: This seems like the perfect opportunity for a `if let config.append(..)` or also I think you could avoid this second branch by using `match Config::load(file).and_then(Config::append) { ... }`.

File: src/main.rs [hunk 3]
```diff
@@ -30,15 +31,11 @@ struct Opts {
...
-        name = "config",
-        value_name = "FILE",
-        short,
-        long,
-        default_value = "/etc/vector/vector.toml"
-    )]
-    config_path: PathBuf,
+    /// Read configuration from one or more files. Wildcard paths are supported.
+    /// If zero files are specified the default config path
+    /// `/etc/vector/vector.toml` will be targeted.
```
[the remark this answers is not in the document]
Comment (Jeffail) [reply src/main.rs]: True, opened https://github.com/timberio/vector/issues/1738

**Pull request merged**